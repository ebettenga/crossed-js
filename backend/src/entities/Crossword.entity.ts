import { Entity, PrimaryGeneratedColumn, Column } from 'typeorm';

export type Clues = {
  across: string[];
  down: string[];
};

@Entity()
export class Crossword {
  @PrimaryGeneratedColumn()
  id: number;

  @Column('jsonb')
  clues: Clues;

  @Column('jsonb')
  answers: {
    across: string[];
    down: string[];
  };

  @Column('text', { nullable: true })
  author: string;

  @Column('text', { nullable: true })
  created_by: string;

  @Column('text', { nullable: true })
  creator_link: string;

  @Column('simple-array', { nullable: true })
  circles: number[];

  @Column('date', { nullable: true })
  date: Date;

  @Column('text', { nullable: true })
  dow: string;

  @Column('simple-array', { nullable: true })
  grid: string[];

  @Column('simple-array', { nullable: true })
  gridnums: string[];

  @Column('boolean', { nullable: true })
  shadecircles: boolean;

  @Column('int', { nullable: true })
  col_size: number;

  @Column('int', { nullable: true })
  row_size: number;

  @Column('text', { nullable: true })
  jnote: string;

  @Column('text', { nullable: true })
  notepad: string;

  @Column('text', { nullable: true })
  title: string;
}
