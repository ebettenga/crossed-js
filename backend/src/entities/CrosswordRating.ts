import {
  Column,
  CreateDateColumn,
  Entity,
  ManyToOne,
  OneToOne,
  PrimaryGeneratedColumn,
} from "typeorm";
import type { User } from "./User";
import type { Crossword } from "./Crossword";

export type DifficultyRating = "too_easy" | "just_right" | "too_hard";

@Entity()
export class CrosswordRating {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne("User", { eager: true })
  user: User;

  @Column({ type: "int" })
  userId: number;

  @ManyToOne("Crossword", { eager: true })
  crossword: Crossword;

  @Column({ type: "int" })
  crosswordId: number;

  @Column({
    type: "enum",
    enum: ["too_easy", "just_right", "too_hard"],
    nullable: true,
  })
  difficultyRating: DifficultyRating;

  @Column({
    type: "int",
    nullable: true,
  })
  qualityRating: number;

  @CreateDateColumn()
  created_at: Date;

  @Column({ type: "timestamp", nullable: true })
  updated_at: Date;

  toJSON() {
    const { user, crossword, ...rest } = this;
    return {
      ...rest,
      user: user.toJSON(),
    };
  }
}
