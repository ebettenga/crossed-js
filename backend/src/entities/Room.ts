import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  JoinTable,
  ManyToMany,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
} from "typeorm";
import type { User } from "./User";
import type { Crossword } from "./Crossword";
import type { GameStats } from "./GameStats";
import type { CachedGameInfo } from "../services/RedisService";

export type GameType = "1v1" | "2v2" | "free4all" | "time_trial";
export type GameStatus = "playing" | "pending" | "finished" | "cancelled";

type Clues = {
  across: string[];
  down: string[];
};

enum PopulatingState {
  READING,
  WRITING,
}

interface Clue {
  number: number;
  hint: string;
}

export enum SquareType {
  SOLVED,
  BLANK,
  BLACK,
  CIRCLE_BLANK,
  CIRCLE_SOLVED,
}

export interface Square {
  id: number;
  squareType: SquareType;
  letter?: string;
  gridnumber: number | null;
  x: number;
  y: number;
  downQuestion?: string;
  acrossQuestion?: string;
}

@Entity()
export class Room {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({
    type: "enum",
    enum: ["1v1", "2v2", "free4all", "time_trial"],
    default: "1v1",
  })
  type: GameType;

  @Column({
    type: "enum",
    enum: ["playing", "pending", "finished", "cancelled"],
    default: "pending",
  })
  status: GameStatus;

  @ManyToMany("User", { eager: true })
  @JoinTable({
    name: "room_players",
    joinColumn: { name: "room_id", referencedColumnName: "id" },
    inverseJoinColumn: { name: "user_id", referencedColumnName: "id" },
  })
  players: User[];

  @ManyToOne("Crossword", { eager: true })
  @JoinColumn()
  crossword: Crossword;

  // Store scores as a JSON object with user IDs as keys
  @Column("simple-json", { default: {} })
  scores: { [key: number]: number };

  @CreateDateColumn()
  created_at: Date;

  @Column({ type: "timestamp", nullable: true })
  completed_at: Date;

  @Column("text")
  difficulty: string;

  @Column("char", { array: true, default: "{}" })
  found_letters: string[];

  // @ts-ignore
  @OneToMany("GameStats", (gameStats) => gameStats.room)
  stats: GameStats[];

  @Column({ type: "timestamp", nullable: true })
  last_activity_at: Date;

  // Cache for toView result
  private viewCache: any = null;
  private lastModified: number = 0;
  private lastViewUpdate: number = 0;

  // Method to mark the room as modified
  markModified() {
    this.lastModified = Date.now();
    this.viewCache = null;
  }

  get player_count(): number {
    return this.players.length;
  }

  createRoomCache(): CachedGameInfo {
    const initialGuessCounts = this.players.reduce((acc, player) => {
      acc[player.id] = {
        correct: 0,
        incorrect: 0,
      };
      return acc;
    }, {} as CachedGameInfo["userGuessCounts"]);

    const initialGuessDetails = this.players.reduce((acc, player) => {
      acc[player.id] = [];
      return acc;
    }, {} as CachedGameInfo["correctGuessDetails"]);

    return {
      lastActivityAt: Date.now(),
      foundLetters: [...this.found_letters],
      scores: { ...this.scores },
      userGuessCounts: initialGuessCounts,
      correctGuessDetails: initialGuessDetails,
    };
  }

  toJSON(foundLetters?: string[], scores?: any): any {
    // If cache exists and room hasn't been modified, return cached view
    if (this.viewCache && this.lastModified === this.lastViewUpdate) {
      return this.viewCache;
    }

    const view = {
      id: this.id,
      created_at: this.created_at,
      completed_at: this.completed_at,
      difficulty: this.difficulty,
      type: this.type,
      status: this.status,
      player_count: this.player_count,
      players: this.players.map((player) => ({
        id: player.id,
        username: player.username,
        score: this.scores[player.id] || 0,
        eloRating: player.eloRating,
      })),
      scores: scores ? scores : this.scores,
      crossword: {
        id: this.crossword.id,
        col_size: this.crossword.col_size,
        row_size: this.crossword.row_size,
        gridnums: this.crossword.gridnums,
        clues: {
          across: this.crossword.clues.across,
          down: this.crossword.clues.down,
        },
        title: this.crossword.title,
        author: this.crossword.author,
        created_by: this.crossword.created_by,
        creator_link: this.crossword.creator_link,
      },
      found_letters: foundLetters ? foundLetters : this.found_letters,
      board: this.createBoard(),
      stats: null,
    };

    // Update cache
    this.viewCache = view;
    this.lastViewUpdate = this.lastModified;

    return view;
  }

  private getSquareType(
    hasLetter: boolean,
    isCircled: boolean,
    black: boolean,
  ): SquareType {
    if (black) return SquareType.BLACK;
    if (hasLetter) {
      return isCircled ? SquareType.CIRCLE_SOLVED : SquareType.SOLVED;
    } else {
      return isCircled ? SquareType.CIRCLE_BLANK : SquareType.BLANK;
    }
  }

  private createSquares(): Square[] {
    return this.found_letters.map((letterCharacter: string, index) => {
      const x = Math.floor(index / this.crossword.row_size);
      return {
        id: index,
        x,
        y: index - x * this.crossword.row_size,
        squareType: this.getSquareType(
          /[a-zA-z]/.test(letterCharacter),
          this.crossword.circles ? this.crossword.circles[index] === 1 : false,
          letterCharacter === ".",
        ),
        gridnumber: parseInt(this.crossword.gridnums[index]) !== 0
          ? parseInt(this.crossword.gridnums[index])
          : null,
        letter: letterCharacter !== "*" ? letterCharacter : undefined,
      } as Square;
    });
  }

  private arrayToMatrix<T>(array: T[], row_length: number): T[][] {
    return Array(Math.ceil(array.length / row_length))
      .fill("")
      .reduce((acc, _, index) => {
        return [...acc, [...array].splice(index * row_length, row_length)];
      }, [] as T[][]);
  }

  private createClueArray(clues: string[]): Clue[] {
    return clues.map((clue) => {
      const clueArray = clue.split(".");
      const clueNumber = parseInt(clueArray[0]);
      return { number: clueNumber, hint: clue };
    });
  }

  private getClueByQuestionNumber(
    clueList: Clue[],
    questionNumber: number | null,
  ): Clue | undefined {
    return clueList.find((clue) => clue.number === questionNumber);
  }

  private createBoard(): Square[][] {
    const squares = this.createSquares();
    const board = this.arrayToMatrix(squares, this.crossword.row_size);

    // Add clues to squares
    const acrossClues = this.createClueArray(this.crossword.clues.across);
    const downClues = this.createClueArray(this.crossword.clues.down);

    // Populate across clues
    board.forEach((row, rowIndex) => {
      row.forEach((square, colIndex) => {
        // Find most recent non-black square in the row or first square
        let lastNonBlackSquare: Square | null = square;
        for (let i = colIndex; i >= 0; i--) {
          if (row[i].squareType !== SquareType.BLACK) {
            lastNonBlackSquare = row[i];
          }
          if (row[i].squareType === SquareType.BLACK) {
            break;
          }
        }

        if (lastNonBlackSquare) {
          square.acrossQuestion = this.getClueByQuestionNumber(
            acrossClues,
            lastNonBlackSquare.gridnumber,
          )?.hint;
        }
      });
    });
    // Populate down clues

    board.forEach((row, rowIndex) => {
      row.forEach((square, colIndex) => {
        // Find most recent non-black square in the row or first square
        let lastNonBlackSquare: Square | null = square;
        for (let i = rowIndex; i >= 0; i--) {
          if (board[i][colIndex].squareType !== SquareType.BLACK) {
            lastNonBlackSquare = board[i][colIndex];
          }
          if (board[i][colIndex].squareType === SquareType.BLACK) {
            break;
          }
        }
        if (lastNonBlackSquare) {
          square.downQuestion = this.getClueByQuestionNumber(
            downClues,
            lastNonBlackSquare.gridnumber,
          )?.hint;
        }
      });
    });

    return board;
  }
}
