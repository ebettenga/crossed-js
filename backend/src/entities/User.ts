import { Entity, PrimaryGeneratedColumn, Column, OneToOne } from 'typeorm';

@Entity()
export class User {
  @PrimaryGeneratedColumn()
  id: number;

  @Column('text')
  username: string;

  @Column('text', { nullable: true })
  githubId?: string;

  @Column('enum', { enum: ['USER', 'ADMIN'], default: 'USER' })
  role: string;

  @Column({ type: 'timestamp', default: () => 'CURRENT_TIMESTAMP' })
  created_at: Date;

  @Column({
    type: 'timestamp',
    default: () => 'CURRENT_TIMESTAMP',
    onUpdate: 'CURRENT_TIMESTAMP',
  })
  updated_at: Date;
}
