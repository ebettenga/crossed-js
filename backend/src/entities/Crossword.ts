
import { Entity, PrimaryGeneratedColumn, Column } from 'typeorm';
import { Room } from './Room';

@Entity()
export class Crossword {
  @PrimaryGeneratedColumn()
  id: number;

  @Column('jsonb')
  clues: object;

  @Column('jsonb')
  answers: object;

  @Column('text', { nullable: true })
  author: string;

  @Column('simple-array', { nullable: true })
  circles: number[];

  @Column('date', { nullable: true })
  date: Date;

  @Column('text', { nullable: true })
  dow: string;

  @Column('simple-array', { nullable: true })
  grid: string[];

  @Column('simple-array', { nullable: true })
  gridnums: number[];

  @Column('boolean', { nullable: true })
  shadecircles: boolean;

  @Column('int', { nullable: true })
  col_size: number;

  @Column('int', { nullable: true })
  row_size: number;

  @Column('text', { nullable: true })
  jnote: string;

  @Column('text', { nullable: true })
  notepad: string;

  @Column('text', { nullable: true })
  title: string;

  room: Room
}
