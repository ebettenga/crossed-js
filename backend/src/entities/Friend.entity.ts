import {
  Column,
  CreateDateColumn,
  Entity,
  ManyToOne,
  PrimaryGeneratedColumn,
} from "typeorm";
import { User } from "./User.entity";

export enum FriendshipStatus {
  PENDING = "pending",
  ACCEPTED = "accepted",
  REJECTED = "rejected",
}

@Entity()
export class Friend {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne(() => User, { eager: true })
  sender: User;

  @ManyToOne(() => User, { eager: true })
  receiver: User;

  @Column("integer")
  senderId: number;

  @Column("integer")
  receiverId: number;

  @Column({
    type: "enum",
    enum: FriendshipStatus,
    default: FriendshipStatus.PENDING,
  })
  status: FriendshipStatus;

  @CreateDateColumn()
  createdAt: Date;

  @Column({ type: "timestamp", nullable: true })
  acceptedAt: Date;
}
