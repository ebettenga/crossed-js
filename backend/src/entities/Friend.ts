import { Entity, PrimaryGeneratedColumn, Column, ManyToOne, CreateDateColumn } from "typeorm";
import { User } from "./User";

export enum FriendshipStatus {
  PENDING = "pending",
  ACCEPTED = "accepted",
  REJECTED = "rejected"
}

@Entity()
export class Friend {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne(() => User, { eager: true })
  sender: User;

  @ManyToOne(() => User, { eager: true })
  receiver: User;

  @Column("integer")
  senderId: number;

  @Column("integer")
  receiverId: number;

  @Column({
    type: "enum",
    enum: FriendshipStatus,
    default: FriendshipStatus.PENDING
  })
  status: FriendshipStatus;

  @CreateDateColumn()
  createdAt: Date;

  @Column({ type: "timestamp", nullable: true })
  acceptedAt: Date;
} 