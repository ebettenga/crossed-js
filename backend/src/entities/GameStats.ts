import { Entity, PrimaryGeneratedColumn, Column, ManyToOne, CreateDateColumn } from "typeorm";
import { User } from "./User";
import { Room } from "./Room";

interface GuessRecord {
  x: number;
  y: number;
  letter: string;
  timestamp: Date;
}

@Entity()
export class GameStats {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne(() => User, { eager: true })
  user: User;

  @ManyToOne(() => Room, { eager: true })
  room: Room;

  @Column("integer")
  userId: number;

  @Column("integer")
  roomId: number;

  @Column("integer")
  correctGuesses: number;

  @Column("integer")
  incorrectGuesses: number;

  @Column("integer")
  userEloAtGame: number;

  @Column("jsonb", { nullable: true })
  correctGuessRecords: GuessRecord[];

  @CreateDateColumn()
  createdAt: Date;

  constructor() {
    this.correctGuesses = 0;
    this.incorrectGuesses = 0;
    this.correctGuessRecords = [];
  }
}