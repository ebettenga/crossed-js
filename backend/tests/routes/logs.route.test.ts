import Fastify from "fastify";
import { DataSource } from "typeorm";
import type { PluginDataSource } from "typeorm-fastify-plugin";
import { ensureApprovedSnapshot } from "../utils/approval";
import { Crossword } from "../../src/entities/Crossword";
import { GameStats } from "../../src/entities/GameStats";
import { Log } from "../../src/entities/Log";
import { Room } from "../../src/entities/Room";
import { User } from "../../src/entities/User";
import logsRoutes from "../../src/routes/private/logs";

jest.setTimeout(60000);

const TEST_DB =
  process.env.LOGS_ROUTE_TEST_DB ||
  process.env.ROOM_SERVICE_TEST_DB ||
  process.env.POSTGRES_DB ||
  "crossed_test";

if (!/_test$/i.test(TEST_DB)) {
  throw new Error(
    `Logs route tests require a test database (received "${TEST_DB}").`,
  );
}

const TEST_SCHEMA =
  process.env.LOGS_ROUTE_TEST_SCHEMA ||
  process.env.ROOM_SERVICE_TEST_SCHEMA ||
  "logs_route_test";

const TEST_HOST =
  process.env.LOGS_ROUTE_TEST_DB_HOST ||
  process.env.ROOM_SERVICE_TEST_DB_HOST ||
  process.env.PGHOST ||
  "127.0.0.1";

const TEST_PORT = parseInt(
  process.env.LOGS_ROUTE_TEST_DB_PORT ||
    process.env.ROOM_SERVICE_TEST_DB_PORT ||
    process.env.PGPORT ||
    "5432",
  10,
);

const TEST_USER =
  process.env.LOGS_ROUTE_TEST_DB_USER ||
  process.env.ROOM_SERVICE_TEST_DB_USER ||
  process.env.PGUSER ||
  "postgres";

const TEST_PASSWORD =
  process.env.LOGS_ROUTE_TEST_DB_PASSWORD ||
  process.env.ROOM_SERVICE_TEST_DB_PASSWORD ||
  process.env.PGPASSWORD ||
  "postgres";

const baseConnectionOptions = {
  type: "postgres" as const,
  host: TEST_HOST,
  port: TEST_PORT,
  username: TEST_USER,
  password: TEST_PASSWORD,
  database: TEST_DB,
};

const qualified = (tableName: string) =>
  `"${TEST_SCHEMA}"."${tableName}"`;

let dataSource: DataSource;

const ensureSchema = async () => {
  const admin = new DataSource({
    ...baseConnectionOptions,
    synchronize: false,
    entities: [],
  });
  await admin.initialize();
  await admin.query(`CREATE SCHEMA IF NOT EXISTS "${TEST_SCHEMA}"`);
  await admin.destroy();
};

const initialiseDataSource = async () => {
  dataSource = new DataSource({
    ...baseConnectionOptions,
    schema: TEST_SCHEMA,
    synchronize: true,
    entities: [Crossword, GameStats, Log, Room, User],
  });
  await dataSource.initialize();
};

const truncateLogs = async () => {
  await dataSource.query(
    `TRUNCATE TABLE ${qualified("log")} RESTART IDENTITY CASCADE`,
  );
};

const buildServer = async (user: User) => {
  const app = Fastify({ logger: false });

  // Match the plugin typing without booting the full Fastify plugin in tests
  app.decorate("orm", dataSource as unknown as PluginDataSource);
  app.decorateRequest("user", null);
  app.addHook("preHandler", (request, _reply, done) => {
    request.user = user;
    done();
  });

  logsRoutes(app as any, {}, () => {});
  await app.ready();
  return app;
};

const createUser = (overrides: Partial<User> = {}) =>
  Object.assign(new User(), {
    id: 101,
    username: "tester",
    email: "tester@example.com",
    password: "secret",
    roles: ["user"],
    status: "online",
    eloRating: 1200,
    ...overrides,
  });

const createLogEntry = async (
  attributes: Partial<Log> & { severity: string; log: object },
) => {
  const repository = dataSource.getRepository(Log);
  const entry = repository.create({
    created_at: new Date("2024-01-01T00:00:00.000Z"),
    ...attributes,
  });
  return repository.save(entry);
};

beforeAll(async () => {
  await ensureSchema();
  await initialiseDataSource();
});

beforeEach(async () => {
  await truncateLogs();
});

afterAll(async () => {
  if (dataSource?.isInitialized) {
    await dataSource.destroy();
  }
});

describe("logs routes", () => {
  it("returns persisted logs matching the snapshot", async () => {
    const user = createUser();
    await createLogEntry({
      severity: "info",
      log: { message: "First entry", userId: 10 },
      created_at: new Date("2024-01-01T00:00:00.000Z"),
    });
    await createLogEntry({
      severity: "error",
      log: { message: "Second entry", userId: 11 },
      created_at: new Date("2024-01-02T00:00:00.000Z"),
    });

    const app = await buildServer(user);
    const response = await app.inject({
      method: "GET",
      url: "/logs",
    });
    expect(response.statusCode).toBe(200);

    const payload = response.json() as Array<{
      id: number;
      created_at: string;
      severity: string;
      log: object;
    }>;

    const sanitized = payload.map((entry) => ({
      id: entry.id,
      created_at: entry.created_at,
      severity: entry.severity,
      log: entry.log,
    }));

    await ensureApprovedSnapshot({
      testFile: expect.getState().testPath ?? "logs.route.test.ts",
      snapshotName: "returns persisted logs matching the snapshot",
      received: sanitized,
    });

    await app.close();
  });

  it("creates a log record and overrides body userId with authenticated user", async () => {
    const user = createUser({ id: 202 });
    const app = await buildServer(user);

    const body = {
      severity: "warn",
      log: { message: "Example log from client", userId: 999 },
    };

    const response = await app.inject({
      method: "POST",
      url: "/logs",
      payload: body,
    });

    expect(response.statusCode).toBe(201);
    const payload = response.json();

    expect(payload.log.userId).toBe(user.id);
    expect(payload.severity).toBe(body.severity);
    expect(payload.id).toBeGreaterThan(0);

    const repository = dataSource.getRepository(Log);
    const stored = await repository.findOneByOrFail({ id: payload.id });

    expect(stored.log).toEqual({
      message: body.log.message,
      userId: user.id,
    });
    expect(stored.severity).toBe(body.severity);

    await app.close();
  });

  it("rejects log creation when severity is missing", async () => {
    const user = createUser();
    const app = await buildServer(user);

    const response = await app.inject({
      method: "POST",
      url: "/logs",
      payload: { log: { message: "No severity" } },
    });

    expect(response.statusCode).toBeGreaterThanOrEqual(400);

    await app.close();
  });
});
