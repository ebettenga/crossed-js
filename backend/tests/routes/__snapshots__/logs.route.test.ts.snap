// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logs routes returns persisted logs matching the snapshot 1`] = `
Array [
  Object {
    "created_at": "2024-01-01T00:00:00.000Z",
    "id": 1,
    "log": Object {
      "message": "First entry",
      "userId": 10,
    },
    "severity": "info",
  },
  Object {
    "created_at": "2024-01-02T00:00:00.000Z",
    "id": 2,
    "log": Object {
      "message": "Second entry",
      "userId": 11,
    },
    "severity": "error",
  },
]
`;
